### PR TITLE
Ingesting tabulate as an internal module

### LINKED ISSUE #1749: Replace `tabulate` with `prettytable`
### Discussed in https://github.com/terrapower/armi/discussions/1747

<div type='discussions-op-text'>

<sup>Originally posted by **opotowsky** June 25, 2024</sup>
`tabulate` is a dead project and `prettytable` is not. But it's pretty extensively used here and downstream with the `runLog` module. </div>

**NOTE**: Because `tabulate` is a dead project, it is stopping us from moving to Python 3.11.

### PR DESCRIPTION
**Pull request by john-science**
## What is the change?

I removed the `tabulate` dependency, and just ingested the parts of that tool we use to `armi.utils.tabulate`.

## Why is the change being made?

To close #1749

The problem is that `tabulate` appears to be a nigh abandoned project, and it is keeping us from being able to support modern PIP versions.

(Arrielle also points out that we could just _replace_ `tabulate` with something else, like `prettytable`. I am not super convinced my approach here is better than that one. I went this route first because it was easy for me to do, and would be easier to make the update for our downstream repos.)

## Code Coverage

The [code coverage](https://coveralls.io/builds/69122857/source?filename=armi%2Futils%2Ftabulate.py) on this file is pretty high:

image

---

## Checklist

- [x] This PR has only [one purpose or idea](https://terrapower.github.io/armi/developer/tooling.html#one-idea-one-pr).
- [x] [Tests](https://terrapower.github.io/armi/developer/tooling.html#test-it) have been added/updated to verify any new/changed code.

- [x] The code style follows [good practices](https://terrapower.github.io/armi/developer/standards_and_practices.html).
- [x] The commit message(s) follow [good practices](https://terrapower.github.io/armi/developer/tooling.html).

- [x] The [release notes](https://terrapower.github.io/armi/developer/tooling.html#add-release-notes) have been updated if necessary.
- [x] The [documentation](https://terrapower.github.io/armi/developer/tooling.html#document-it) is still up-to-date in the `doc` folder.
- [x] The dependencies are still up-to-date in `pyproject.toml`.

### CODE REVIEW

File: armi/utils/tabulate.py [hunk 1]
```diff
@@ -0,0 +1,1583 @@
...
+    missingval=_DEFAULT_MISSINGVAL,
+    showindex="default",
+    disableNumparse=False,
+    colglobalalign=None,
+    colalign=None,
+    maxcolwidths=None,
+    headersglobalalign=None,
+    headersalign=None,
+    rowalign=None,
+    maxheadercolwidths=None,
```
Comment (opotowsky) [inline armi/utils/tabulate.py:1026]: how much work is it to get these standardized to ARMI's style guide (🐫)

File: armi/utils/tabulate.py [hunk 2]
```diff
@@ -0,0 +1,1583 @@
...
+):
+    colwidths = [w - 2 * pad for w in paddedWidths]
+    cells_lines = [c.splitlines() for c in paddedMultilineCells]
+    nlines = max(map(len, cells_lines))  # number of lines in the row
+
+    cells_lines = [
+        _alignCellVeritically(cl, nlines, w, rowalign)
+        for cl, w in zip(cells_lines, colwidths)
+    ]
+    linesCells = [[cl[i] for cl in cells_lines] for i in range(nlines)]
```
Comment (opotowsky) [inline armi/utils/tabulate.py:1494]: 🐍 case -> 🐫 case

File: armi/utils/tabulate.py (hunk 1 above)
Comment (opotowsky) [reply armi/utils/tabulate.py:1026]: A lot are just all lower case and could be camel, but I won't die on that hill!

File: armi/utils/tabulate.py [hunk 3]
```diff
@@ -0,0 +1,1583 @@
...
+    "'42'"
+
+    """
+    if isinstance(s, bytes):
+        return s.decode(encoding=encoding, errors=errors)
+    return str(s)
+
+
+def tabulate(
+    tabular_data,
```
Comment (opotowsky) [inline armi/utils/tabulate.py:1010]: 🐍 case -> 🐫 case

File: armi/utils/tests/test_tabulate.py [hunk 4]
```diff
@@ -0,0 +1,1697 @@
...
+class TestTabulateOutput(unittest.TestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.test_table = [["spam", 41.9999], ["eggs", "451.0"]]
+        cls.test_table_with_sep_line = [
+            ["spam", 41.9999],
+            SEPARATING_LINE,
+            ["eggs", "451.0"],
+        ]
+        cls.test_table_headers = ["strings", "numbers"]
```
Comment (opotowsky) [inline armi/utils/tests/test_tabulate.py:655]: 🐍 case -> 🐫 case

File: armi/utils/tests/test_tabulate.py [hunk 5]
```diff
@@ -0,0 +1,1697 @@
...
+from armi.utils.tabulate import _wrapTextToColWidths
+from armi.utils.tabulate import SEPARATING_LINE
+from armi.utils.tabulate import tabulate
+from armi.utils.tabulate import tabulate_formats
+
+
+class TestTabulateAPI(unittest.TestCase):
+    def test_tabulateFormats(self):
+        """API: tabulate_formats is a list of strings."""
+        supported = tabulate_formats
```
Comment (opotowsky) [inline armi/utils/tests/test_tabulate.py:52]: 🐍 case -> 🐫 case

File: armi/utils/tabulate.py [hunk 6]
```diff
@@ -0,0 +1,1583 @@
...
+    ),
+}
+
+
+tabulate_formats = list(sorted(_table_formats.keys()))
+
+# The table formats for which multiline cells will be folded into subsequent table rows. The key is
+# the original format specified at the API. The value is the format that will be used to represent
+# the original format.
+multiline_formats = {
```
Comment (opotowsky) [inline armi/utils/tabulate.py:228]: 🐍 case -> 🐫 case

**Review by opotowsky — COMMENTED** [review]
I agree that ingesting the file should be less work overall. 

I think there's a few more places you could scrape for 🐍 case -> 🐫 case.

GREAT JOB! Thanks so much!

File: armi/utils/tabulate.py (hunk 3 above)
Comment (john-science) [reply armi/utils/tabulate.py:1010]: The rest I agree with you on. 

Unnnnnfortunately...

This one this one and `tabulate_formats` below are already exposed as `snake_case` in the API.  And I checked, we have dozens of ARMI Plugins downstream that explicitly call these two out, and would need to be changed. 

THIS version of the PR means nothing downstream breaks with this PR, it's just a import change.

#shrug

File: armi/utils/tabulate.py (hunk 6 above)
Comment (john-science) [reply armi/utils/tabulate.py:228]: The rest I agree with you on. 

Unnnnnfortunately...

This two and `tabular_data` above are already exposed as `snake_case` in the API.  And I checked, we have dozens of ARMI Plugins downstream that explicitly call these two out, and would need to be changed. 

THIS version of the PR means nothing downstream breaks with this PR, it's just a import change.

#shrug

File: armi/utils/tabulate.py (hunk 2 above)
Comment (john-science) [reply armi/utils/tabulate.py:1494]: ✅

File: armi/utils/tests/test_tabulate.py (hunk 4 above)
Comment (john-science) [reply armi/utils/tests/test_tabulate.py:655]: ✅

File: armi/utils/tabulate.py (hunk 1 above)
Comment (john-science) [reply armi/utils/tabulate.py]: Done!

File: armi/utils/tests/test_tabulate.py (hunk 5 above)
Comment (john-science) [reply armi/utils/tests/test_tabulate.py:52]: The rest I agree with you on. 

Unnnnnfortunately...

There are dozens of downstream repositories that already use three `snake_case` pieces of the `tabulate` API in hundreds of unique lines:

* `tabulate_formats`
* `multiline_formates`
* `tabular_data`

So, this is my attempt to keep the downstream changes after this PR down to the bare minimum of `import` changes only.

Thoughts?

File: armi/utils/tabulate.py [hunk 7]
```diff
@@ -41,7 +41,7 @@
 
 _DEFAULT_FLOATFMT = "g"
 _DEFAULT_INTFMT = ""
-_DEFAULT_MISSINGVAL = ""
+_DEFAULT_missingVal = ""
```
Comment (jakehader) [inline armi/utils/tabulate.py:44]: Find / replace issue

**Comment by jakehader** [comment]
I think it's fine to pull in 🐫 case throughout and then move code over to armi's tabulate as it makes sense. Should be pretty easy to see where to replace these things. Could also replace things that are 🐍 case with a simple name (`tabular_data` -> `data` or something). Should we pull in additional documentation of this new feature into the ARMI docs?

File: armi/utils/tabulate.py (hunk 7 above)
Comment (john-science) [reply armi/utils/tabulate.py]: Fixed!

File: armi/utils/tabulate.py (hunk 6 above)
Comment (john-science) [reply armi/utils/tabulate.py:228]: Okay, based on your opinions and @jakehader , I will just make these API changes, and make everything `camelCase`. NBD.

File: armi/utils/tests/test_tabulate.py (hunk 5 above)
Comment (john-science) [reply armi/utils/tests/test_tabulate.py:52]: Okay, based on your opinions and @jakehader , I will just make these API changes, and make everything `camelCase`. NBD.

File: armi/utils/tabulate.py (hunk 3 above)
Comment (john-science) [reply armi/utils/tabulate.py:1010]: Okay, based on your opinions and @jakehader , I will just make these API changes, and make everything `camelCase`. NBD.

**Comment by john-science** [comment]
> I think it's fine to pull in 🐫 case throughout and then move code over to armi's tabulate as it makes sense. Should be pretty easy to see where to replace these things. Could also replace things that are 🐍 case with a simple name (`tabular_data` -> `data` or something). Should we pull in additional documentation of this new feature into the ARMI docs? For example pull in examples from https://github.com/astanin/python-tabulate?tab=readme-ov-file#library-usage?

Most of that is already included in the docstring for the `tabulate.tabulate()` function:

https://github.com/terrapower/armi/blob/cac33246db511304058a50837fa24349b13fcb78/armi/utils/tabulate.py#L1027-L1043

I considered putting what you linked as a module-level docstring, and reducing this ridiculously long function docstring (>200 lines). I could do that, and find any missing pieces from the docs you linked as well.

**Review by opotowsky — APPROVED** [review]
We are waiting to merge, but in the meantime: approved! `tabulate` is officially pac-manned!

**Comment by john-science** [comment]
@jakehader I grabbed that README and put it in the module-level docstring in this file.  Thanks!

**Review by jakehader — APPROVED** [review]

**Pull request merged**